include/libopencm3/stm32/f4/ltdc.h

```diff
@@ -490,12 +490,20 @@ static inline bool LTDC_SRCR_IS_RELOADING(void)
  * color conversion helper function
  * (simulate the ltdc color conversion)
  */
+static inline uint32_t ltdc_get_rgb888_from_rgb565(uint16_t rgb565)
+{
+	return   (uint32_t)(((rgb565 & 0xF800) >> (11-8))/31)<<16
+	       | (uint32_t)(((rgb565 & 0x07E0) <<  (8-5))/63)<<8
+	       | (uint32_t)(((rgb565 & 0x001F) <<  (8-0))/31)<<0;
+}
 
-static inline uint16_t ltdc_get_rgb888_from_rgb565(uint16_t rgb888)
+static inline uint16_t ltdc_get_rgb565_from_rgb888(uint32_t rgb888)
 {
-	return ((((rgb888) & 0xF800)   >> (11-8))/31)<<16
-	       | ((((rgb888) & 0x07E0) <<  (8-5))/63)<<8
-	       | ((((rgb888) & 0x001F) <<  (8-0))/31)<<0;
+	return (uint16_t) (
+			 (((((rgb888 & 0xFF0000) >> 16)*31) << (11-8)) & 0xF800)
+	       | (((((rgb888 & 0x00FF00) >>  8)*63) >> (8-5) ) & 0x07E0)
+	       | (((((rgb888 & 0x0000FF) >>  0)*31) >> (8-0) ) & 0x001F)
```

Comment:
Seems a bit like you are moving the bits more than needed, have you tried something like
```
    return (uint16_t)(
            ((rgb888 & 0xf80000) >> 16)  | /* grab upper 5 bits of red)
            ((rgb888 & 0xfc00) >> 5) | /* grab 6 bits of green */
            ((rgb888 & 0xf8) >> 3); /* grab 5 bits of blue */
    );
```